### PR TITLE
Reservation branch

### PR DESCRIPTION
**Pull request by miccu64**
2 things I were unsure:
1. field is_checked will be enough for approving reservation?
2. reservation user should be related with users from USOS from Activity class?

### CODE REVIEW

File: backend/src/main/java/com/wpz/rbs/controller/ReservationController.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+public class ReservationController {
+    private final ReservationService reservationService;
+
+    public ReservationController(ReservationService reservationService) {
+        this.reservationService = reservationService;
+    }
+
+    @PostMapping()
+    public int createReservation(@RequestBody @Validated CreateReservation reservation) throws Exception {
+        return this.reservationService.createReservation(reservation);
```
Comment (bestemic) [inline backend/src/main/java/com/wpz/rbs/controller/ReservationController.java:21]: without this

File: backend/src/main/java/com/wpz/rbs/model/reservation/CreateReservation.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    private String email;
+    @NotEmpty
+    private String first_name;
+    @NotEmpty
+    private String last_name;
+    @NotEmpty
+    private String start_time;
+    @NotEmpty
+    private String end_time;
+    private int room_id;
```
Comment (bestemic) [inline backend/src/main/java/com/wpz/rbs/model/reservation/CreateReservation.java:26]: can this be empty?

**Review by bestemic — CHANGES_REQUESTED** [review]

File: backend/src/main/java/com/wpz/rbs/model/reservation/CreateReservation.java (hunk 2 above)
Comment (miccu64) [reply backend/src/main/java/com/wpz/rbs/model/reservation/CreateReservation.java:26]: Yes, bcs int isn't nullable and throws error when not provided, but I will add @NotEmpty for clearness

File: backend/src/main/java/com/wpz/rbs/controller/ReservationController.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        return reservationService.getById(id);
+    }
+
+    @GetMapping("room/{roomId}")
+    public List<Reservation> getReservationsByRoom(@PathVariable int roomId) {
+        return reservationService.getByRoomId(roomId);
+    }
+
+    @GetMapping()
+    public List<Reservation> getAllReservations() {
```
Comment (bestemic) [inline backend/src/main/java/com/wpz/rbs/controller/ReservationController.java:39]: This isn't needed

File: backend/src/main/java/com/wpz/rbs/controller/ReservationController.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        return reservationService.createReservation(reservation);
+    }
+
+    @GetMapping("{id}")
+    public Reservation getReservationById(@PathVariable int id) {
+        return reservationService.getById(id);
+    }
+
+    @GetMapping("room/{roomId}")
+    public List<Reservation> getReservationsByRoom(@PathVariable int roomId) {
```
Comment (bestemic) [inline backend/src/main/java/com/wpz/rbs/controller/ReservationController.java:34]: This isn't needed

File: backend/src/main/java/com/wpz/rbs/controller/ReservationController.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    @PostMapping()
+    public ResponseEntity<?> createReservation(@RequestBody @Validated CreateReservation reservation) throws ParseException {
+        return reservationService.createReservation(reservation);
+    }
+
+    @GetMapping("{id}")
+    public Reservation getReservationById(@PathVariable int id) {
+        return reservationService.getById(id);
```
Comment (bestemic) [inline backend/src/main/java/com/wpz/rbs/controller/ReservationController.java:30]: Should return 404 when no reservation found (now it't 500 and error in application)

File: backend/src/main/java/com/wpz/rbs/service/ReservationService.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    private final RoomRepository roomRepository;
+
+    public ReservationService(ReservationRepository reservationRepository, ActivityRepository activityRepository, RoomRepository roomRepository) {
+        this.reservationRepository = reservationRepository;
+        this.activityRepository = activityRepository;
+        this.roomRepository = roomRepository;
+    }
+
+    public Reservation getById(int id) {
+        return reservationRepository.findById(id).get();
```
Comment (bestemic) [inline backend/src/main/java/com/wpz/rbs/service/ReservationService.java:29]: Now error when no reservation with given id

File: backend/src/main/java/com/wpz/rbs/service/ReservationService.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    public List<Reservation> getByRoomId(int roomId) {
```
Comment (bestemic) [inline backend/src/main/java/com/wpz/rbs/service/ReservationService.java:32]: This isn't needed

File: backend/src/main/java/com/wpz/rbs/service/ReservationService.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        return reservationRepository.findAllByRoom_Id(roomId);
+    }
+
+    public List<Reservation> getAll() {
```
Comment (bestemic) [inline backend/src/main/java/com/wpz/rbs/service/ReservationService.java:36]: This isn't needed

File: backend/src/main/java/com/wpz/rbs/repository/ReservationRepository.java [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+import com.wpz.rbs.model.Reservation;
+import lombok.NonNull;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.CrudRepository;
+
+import java.util.List;
+
+public interface ReservationRepository extends CrudRepository<Reservation, Integer> {
+    @Query(value = "select * from Reservation r where r.room_id = :roomId", nativeQuery = true)
+    List<Reservation> findAllByRoom_Id(int roomId);
```
Comment (bestemic) [inline backend/src/main/java/com/wpz/rbs/repository/ReservationRepository.java:12]: This isn't needed

File: backend/src/main/java/com/wpz/rbs/repository/ReservationRepository.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+    @NonNull List<Reservation> findAll();
```
Comment (bestemic) [inline backend/src/main/java/com/wpz/rbs/repository/ReservationRepository.java:14]: This isn't needed

File: backend/src/main/java/com/wpz/rbs/service/ReservationService.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        if (collisionExists(reservation.getRoom_id(), startDate, endDate))
+            throw new RuntimeException("Room reservation collision detected");
+        return ResponseEntity.ok(reservationRepository.save(new Reservation(reservation)).getId());
+    }
+
+    private boolean collisionExists(int roomId, Date newStartDate, Date newEndDate) {
+        return activityRepository.findAllByRoom_Id(roomId).stream().anyMatch(a -> compareDates(newStartDate, newEndDate, a.getStart_time(), a.getEnd_time()));
+    }
+
+    private boolean compareDates(Date newStartDate, Date newEndDate, String startDateString, String endDateString) {
```
Comment (bestemic) [inline backend/src/main/java/com/wpz/rbs/service/ReservationService.java:56]: We can extract this to our utils class

File: backend/src/main/java/com/wpz/rbs/service/ReservationService.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    public ResponseEntity<?> createReservation(CreateReservation reservation) throws ParseException {
+        Date startDate = StaticHelpers.parseDateTime(reservation.getStart_time());
+        Date endDate = StaticHelpers.parseDateTime(reservation.getEnd_time());
+        if (!roomRepository.existsById(reservation.getRoom_id()))
+            return ResponseEntity.status(400).body("Wrong room id");
+        if (startDate.compareTo(endDate) >= 0)
+            return ResponseEntity.status(409).body("Start date is later than end date");
+        if (collisionExists(reservation.getRoom_id(), startDate, endDate))
+            throw new RuntimeException("Room reservation collision detected");
+        return ResponseEntity.ok(reservationRepository.save(new Reservation(reservation)).getId());
```
Comment (bestemic) [inline backend/src/main/java/com/wpz/rbs/service/ReservationService.java:49]: Post should return created object, not id.

File: backend/src/main/java/com/wpz/rbs/service/ReservationService.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    private boolean collisionExists(int roomId, Date newStartDate, Date newEndDate) {
```
Comment (bestemic) [inline backend/src/main/java/com/wpz/rbs/service/ReservationService.java:52]: Can we do this in sql on database. Will be faster than returning all activities.

File: backend/src/main/java/com/wpz/rbs/utils/StaticHelpers.java [hunk 14]
```diff
@@ -0,0 +1,26 @@
+package com.wpz.rbs.utils;
+
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Calendar;
+import java.util.Date;
+
+public class StaticHelpers {
```
Comment (bestemic) [inline backend/src/main/java/com/wpz/rbs/utils/StaticHelpers.java:8]: Rename to DateUtils

File: backend/src/main/java/com/wpz/rbs/service/ActivityService.java [hunk 15]
```diff
@@ -26,6 +29,19 @@ public List<Activity> getByRoomId(int roomId) {
         return new ArrayList<>(activityRepository.findAllByRoom_Id(roomId));
     }
 
+    public List<Activity> getByRoomIdForNextWeek(int roomId, String startDateString) throws ParseException {
+        Date startDate = StaticHelpers.parseDate(startDateString);
+        Date endDate = StaticHelpers.addDays(startDate, 6);
+        return activityRepository.findAllByRoom_Id(roomId).stream().filter(a -> {
```
Comment (bestemic) [inline backend/src/main/java/com/wpz/rbs/service/ActivityService.java:35]: Can we also do this using query in the db not retrieving all data?

**Review by bestemic — CHANGES_REQUESTED** [review]

**Comment by bestemic** [comment]
Kilka małych zmian, kilka większych które usprawnią działanie.

File: backend/src/main/java/com/wpz/rbs/service/ReservationService.java (hunk 6 above)
Comment (miccu64) [reply backend/src/main/java/com/wpz/rbs/service/ReservationService.java:29]: Corrected

File: backend/src/main/java/com/wpz/rbs/service/ReservationService.java (hunk 11 above)
Comment (miccu64) [reply backend/src/main/java/com/wpz/rbs/service/ReservationService.java:56]: Improved and exported

**Pull request merged**